libgramtools/src/kmer_index/build.cpp

```diff
@@ -39,68 +39,45 @@ CacheElement get_initial_cache_element(const Base &base,
 
-    bool kmer_base_is_last = true;
+    bool kmer_base_is_first_processed = true;
     const auto &cache_element = get_next_cache_element(base,
-                                                       kmer_base_is_last,
-                                                       initial_cache_element,
+                                                       kmer_base_is_first_processed,
+                                                       full_SA_interval,
                                                        prg_info);
     return cache_element;
 }
 
 
-KmerIndexCache initial_kmer_index_cache(const Pattern &full_kmer,
-                                        const PRG_Info &prg_info) {
-    KmerIndexCache cache;
+void build_kmer_cache(KmerIndexCache &cache,
+                      const Pattern &kmer_prefix_diff,
+                      const int kmer_size,
+                      const PRG_Info &prg_info) {
+    auto it = kmer_prefix_diff.rbegin();
 
-    for (auto it = full_kmer.rbegin(); it != full_kmer.rend(); ++it) {
+    if (kmer_prefix_diff.size() == kmer_size) {
         const auto &base = *it;
-        const bool kmer_base_is_last = it == full_kmer.rbegin();
-
-        if (cache.empty()) {
-            auto cache_element = get_initial_cache_element(base, prg_info);
-            cache.emplace_back(cache_element);
-            continue;
-        }
-
-        const auto &last_cache_element = cache.back();
-        if (last_cache_element.search_states.empty()) {
-            cache.emplace_back(CacheElement{});
-            continue;
-        }
-
-        const auto &new_cache_element = get_next_cache_element(base,
-                                                               kmer_base_is_last,
-                                                               last_cache_element,
-                                                               prg_info);
-        cache.emplace_back(new_cache_element);
+        cache.resize(0); // Empty the `cache`
+        auto cache_element = get_initial_cache_element(base, prg_info);
+        cache.emplace_back(cache_element);
+        ++it;
     }
-    return cache;
-}
 
-
-void update_kmer_index_cache(KmerIndexCache &cache,
-                             const Pattern &kmer_prefix_diff,
-                             const int kmer_size,
-                             const PRG_Info &prg_info) {
-    if (kmer_prefix_diff.size() == kmer_size) {
-        auto &full_kmer = kmer_prefix_diff;
-        cache = initial_kmer_index_cache(full_kmer, prg_info);
-        return;
+    else {
+        const auto preserved_cache_size = kmer_size - kmer_prefix_diff.size();
+        cache.resize(preserved_cache_size);
     }
 
-    const auto truncated_cache_size = kmer_size - kmer_prefix_diff.size();
-    cache.resize(truncated_cache_size);
 
-    for (auto it = kmer_prefix_diff.rbegin(); it != kmer_prefix_diff.rend(); ++it) {
+    for (it ; it != kmer_prefix_diff.rend(); ++it) {
```

Comment:
You don't need the first `it` here. This line could be: `for (; it != kmer_prefix_diff.rend(); ++it) {`.